File: languages/Zig/bench.zig
const std = @import("std");
const c = @cImport({
    @cInclude("time.h");
});

const PRIMES_LIMIT: u32 = 20_000_000;
const FIBONACCI_N: u32 = 45;
const STRING_ITER: u32 = 5_000_000;
const MANDEL_W: usize = 4096;
const MANDEL_H: usize = 4096;
const MANDEL_MAX_ITER: i32 = 256;
const TREE_MIN_DEPTH: i32 = 4;
const TREE_MAX_DEPTH: i32 = 18;
const SENTENCE = "The quick brown fox jumps over dat lazy dog that was not enough to jump over the frog again";

fn nowSec() f64 {
    var ts: c.timespec = undefined;
    _ = c.clock_gettime(c.CLOCK_MONOTONIC, &ts);
    return @as(f64, @floatFromInt(ts.tv_sec)) + @as(f64, @floatFromInt(ts.tv_nsec)) / 1e9;
}

fn isPrime(n: u32) bool {
    if (n <= 1) return false;
    if (n <= 3) return true;
    if (n % 2 == 0 or n % 3 == 0) return false;
    var i: u32 = 5;
    while (i * i <= n) : (i += 6) {
        if (n % i == 0 or n % (i + 2) == 0) return false;
    }
    return true;
}

fn benchPrimes() void {
    const out = std.debug.print;
    const start = nowSec();
    var c_count: u32 = 0;
    var i: u32 = 2;
    while (i < PRIMES_LIMIT) : (i += 1) if (isPrime(i)) { c_count += 1; };
    out("Found {} primes in {d:.3} seconds\n\n", .{ c_count, nowSec() - start });
}

fn fib(n: u32) u64 { return if (n <= 1) @as(u64, n) else fib(n - 1) + fib(n - 2); }
fn benchFibRec() void {
    const out = std.debug.print;
    const start = nowSec();
    const r = fib(FIBONACCI_N);
    out("Fibonacci({}) = {} in {d:.3} seconds\n\n", .{ FIBONACCI_N, r, nowSec() - start });
}

fn benchStrings() void {
    const out = std.debug.print;
    const start = nowSec();

    var words: [32][]const u8 = undefined;
    var wcount: usize = 0;
    var iter = std.mem.tokenizeScalar(u8, SENTENCE, ' ');
    while (iter.next()) |w| {
        words[wcount] = w;
        wcount += 1;
    }

    var total: u64 = 0;
    var k: u32 = 0;
    while (k < STRING_ITER) : (k += 1) {
        var i: usize = 0;
        while (i < wcount) : (i += 1) {
            var count: u64 = 0;
            var j: usize = 0;
            while (j < wcount) : (j += 1) {
                if (std.mem.eql(u8, words[i], words[j])) count += 1;
            }
            total += count;
        }
    }
    out("Strings: total={} in {d:.3} seconds\n\n", .{ total, nowSec() - start });
}

fn benchMandelbrot() void {
    const out = std.debug.print;
    const start = nowSec();
    var checksum: i64 = 0;
    var py: usize = 0;
    while (py < MANDEL_H) : (py += 1) {
        const cy: f64 = (@as(f64, @floatFromInt(py)) / @as(f64, @floatFromInt(MANDEL_H))) * 3.0 - 1.5;
        var px: usize = 0;
        while (px < MANDEL_W) : (px += 1) {
            const cx: f64 = (@as(f64, @floatFromInt(px)) / @as(f64, @floatFromInt(MANDEL_W))) * 3.0 - 2.0;
            var zx: f64 = 0.0;
            var zy: f64 = 0.0;
            var iter: i32 = 0;
            while (iter < MANDEL_MAX_ITER and zx * zx + zy * zy <= 4.0) : (iter += 1) {
                const nx = zx * zx - zy * zy + cx;
                zy = 2.0 * zx * zy + cy;
                zx = nx;
            }
            checksum += @as(i64, iter);
        }
    }
    out("Mandelbrot: checksum={} in {d:.3} seconds\n\n", .{ checksum, nowSec() - start });
}

const TreeNode = struct {
    left: ?*TreeNode,
    right: ?*TreeNode,
    item: i64,
};

fn makeTree(allocator: std.mem.Allocator, item: i64, depth: i32) !*TreeNode {
    const n = try allocator.create(TreeNode);
    if (depth == 0) {
        n.* = .{ .left = null, .right = null, .item = item };
        return n;
    }
    n.* = .{
        .left = try makeTree(allocator, 2 * item - 1, depth - 1),
        .right = try makeTree(allocator, 2 * item, depth - 1),
        .item = item,
    };
    return n;
}

fn checkTree(n: *TreeNode) i64 {
    if (n.left == null) return n.item;
    return n.item + checkTree(n.left.?) - checkTree(n.right.?);
}

fn freeTree(allocator: std.mem.Allocator, n: *TreeNode) void {
    if (n.left) |l| {
        freeTree(allocator, l);
        freeTree(allocator, n.right.?);
    }
    allocator.destroy(n);
}

fn benchBinaryTrees() !void {
    const out = std.debug.print;
    const start = nowSec();
    const allocator = std.heap.c_allocator;
    var checksum: i64 = 0;

    {
        const stretch = try makeTree(allocator, 0, TREE_MAX_DEPTH + 1);
        checksum += checkTree(stretch);
        freeTree(allocator, stretch);
    }

    const long_lived = try makeTree(allocator, 0, TREE_MAX_DEPTH);

    var d: i32 = TREE_MIN_DEPTH;
    while (d <= TREE_MAX_DEPTH) : (d += 2) {
        const iters: i32 = @as(i32, 1) << @intCast(TREE_MAX_DEPTH - d + TREE_MIN_DEPTH);
        var sum: i64 = 0;
        var i: i32 = 0;
        while (i < iters) : (i += 1) {
            const t = try makeTree(allocator, @as(i64, i + 1), d);
            sum += checkTree(t);
            freeTree(allocator, t);
        }
        checksum += sum;
    }

    checksum += checkTree(long_lived);
    freeTree(allocator, long_lived);

    out("BinaryTrees: checksum={} in {d:.3} seconds\n\n", .{ checksum, nowSec() - start });
}

pub fn main() !void {
    const out = std.debug.print;
    out("=== PROGRAMMING LANGUAGE BENCHMARK ===\n\n", .{});

    const start = nowSec();
    benchPrimes();
    benchFibRec();
    benchStrings();
    benchMandelbrot();
    try benchBinaryTrees();
    out("=== BENCHMARK COMPLETE ===\n", .{});
    out("Total execution time: {d:.3} seconds\n", .{nowSec() - start});
}

File: languages/Php/bench.php
<?php
const PRIMES_LIMIT = 20000000;
const FIBONACCI_N = 45;
const STRING_ITER = 5000000;
const MANDEL_W = 4096;
const MANDEL_H = 4096;
const MANDEL_MAX_ITER = 256;
const TREE_MIN_DEPTH = 4;
const TREE_MAX_DEPTH = 18;
const SENTENCE = "The quick brown fox jumps over dat lazy dog that was not enough to jump over the frog again";

function is_prime($n) {
    if ($n <= 1) return false;
    if ($n <= 3) return true;
    if ($n % 2 == 0 || $n % 3 == 0) return false;
    for ($i = 5; $i * $i <= $n; $i += 6)
        if ($n % $i == 0 || $n % ($i + 2) == 0) return false;
    return true;
}
function bench_primes() {
    $s = microtime(true);
    $c = 0;
    for ($i = 2; $i < PRIMES_LIMIT; $i++) if (is_prime($i)) $c++;
    printf("Found %d primes in %.3f seconds\n\n", $c, microtime(true) - $s);
}

function fib($n) { return $n <= 1 ? $n : fib($n - 1) + fib($n - 2); }
function bench_fib_rec() {
    $s = microtime(true);
    $r = fib(FIBONACCI_N);
    printf("Fibonacci(%d) = %d in %.3f seconds\n\n", FIBONACCI_N, $r, microtime(true) - $s);
}

function bench_strings() {
    $s = microtime(true);
    $words = explode(' ', SENTENCE);
    $wcount = count($words);
    $total = 0;
    for ($iter = 0; $iter < STRING_ITER; $iter++) {
        for ($i = 0; $i < $wcount; $i++) {
            $count = 0;
            $wi = $words[$i];
            for ($j = 0; $j < $wcount; $j++) {
                if ($wi === $words[$j]) $count++;
            }
            $total += $count;
        }
    }
    printf("Strings: total=%d in %.3f seconds\n\n", $total, microtime(true) - $s);
}

function bench_mandelbrot() {
    $s = microtime(true);
    $checksum = 0;
    for ($py = 0; $py < MANDEL_H; $py++) {
        $cy = ($py / MANDEL_H) * 3.0 - 1.5;
        for ($px = 0; $px < MANDEL_W; $px++) {
            $cx = ($px / MANDEL_W) * 3.0 - 2.0;
            $zx = 0.0; $zy = 0.0;
            $iter = 0;
            while ($iter < MANDEL_MAX_ITER && $zx * $zx + $zy * $zy <= 4.0) {
                $nx = $zx * $zx - $zy * $zy + $cx;
                $zy = 2.0 * $zx * $zy + $cy;
                $zx = $nx;
                $iter++;
            }
            $checksum += $iter;
        }
    }
    printf("Mandelbrot: checksum=%d in %.3f seconds\n\n", $checksum, microtime(true) - $s);
}

class TreeNode {
    public $left;
    public $right;
    public $item;
}
function make_tree($item, $depth) {
    $n = new TreeNode();
    $n->item = $item;
    if ($depth == 0) {
        $n->left = null;
        $n->right = null;
        return $n;
    }
    $n->left = make_tree(2 * $item - 1, $depth - 1);
    $n->right = make_tree(2 * $item, $depth - 1);
    return $n;
}
function check_tree($n) {
    if ($n->left === null) return $n->item;
    return $n->item + check_tree($n->left) - check_tree($n->right);
}
function bench_binary_trees() {
    $s = microtime(true);
    $checksum = 0;

    $stretch = make_tree(0, TREE_MAX_DEPTH + 1);
    $checksum += check_tree($stretch);
    unset($stretch);

    $long_lived = make_tree(0, TREE_MAX_DEPTH);

    for ($d = TREE_MIN_DEPTH; $d <= TREE_MAX_DEPTH; $d += 2) {
        $iters = 1 << (TREE_MAX_DEPTH - $d + TREE_MIN_DEPTH);
        $sum = 0;
        for ($i = 0; $i < $iters; $i++) {
            $t = make_tree($i + 1, $d);
            $sum += check_tree($t);
        }
        $checksum += $sum;
    }

    $checksum += check_tree($long_lived);

    printf("BinaryTrees: checksum=%d in %.3f seconds\n\n", $checksum, microtime(true) - $s);
}

echo "=== PROGRAMMING LANGUAGE BENCHMARK ===\n\n";
$total = microtime(true);
bench_primes();
bench_fib_rec();
bench_strings();
bench_mandelbrot();
bench_binary_trees();
echo "=== BENCHMARK COMPLETE ===\n";
printf("Total execution time: %.3f seconds\n", microtime(true) - $total);
